### PR TITLE
Full js-like member access notation: ${foo.bar[foo.xyzzy[0]]}

### PR DESCRIPTION
**Pull request by iridiankin**
Introduces a string replacement member notation which takes place wherever currently the %%/@@ is used. Notation triggers areas surrounded as ${contents}, where contents is evaluated as js-like field member access as follows:
${globalVariable1Name[0].memberField[globalVariableName2.otherField]}
Only notable diversion from js is that []-index values (for example globalVariableName2.otherField above) that are coerceable to numbers will be coerced regardless of source.
global variable names are looked up against gamestates.
Assignment is not supported (yet).

### CODE REVIEW

File: app/js/errantry/Game.js [hunk 1]
```diff
@@ -5038,6 +5048,61 @@ extObj.height = textObj.origHeight;
     }
     return this.replaceStringsWithValues2(output);
   },
+  parseExpression: function (accessText) {
+    var asNumber = Number(accessText);
+    if (!isNaN(asNumber)) return asNumber;
```
Comment (shundread) [inline app/js/errantry/Game.js]: isNaN() seems to be ECMAScript6

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/isNaN

File: app/js/errantry/Game.js [hunk 2]
```diff
@@ -5018,6 +5018,12 @@ extObj.height = textObj.origHeight;
     var indexes = [];
     for (var i = 0; i < output.length; i++) { // Get indexes of % characters, marking parts to replace with gamestate values
       if (output[i] == '%') indexes.push(i);
+      if (output.slice(i, i + 2) === "${")) {
```
Comment (shundread) [inline app/js/errantry/Game.js]: Mismatching parentheses, should be `if (output.slice(i, i + 2) === "${") {` instead.

File: app/js/errantry/Game.js [hunk 3]
```diff
@@ -5028,6 +5034,10 @@ extObj.height = textObj.origHeight;
       }
 
       for (var i = 0; i < replaceables.length; i++) { // Replace strings with corresponding gamestates
+        if (replaceables[i][0] == "$") {
+          output = output.replace(replaceables[i],
+              parseExpression(replaceables[0].substr(2, replaceables[i].length - 3)));
```
Comment (shundread) [inline app/js/errantry/Game.js]: `this.parseExpression(...)`

File: app/js/errantry/Game.js [hunk 4]
```diff
@@ -5038,6 +5048,61 @@ extObj.height = textObj.origHeight;
...
+    }
+    if ((fieldName == null) || (typeof fieldName == "undefined")) {
+      console.warning("While parsing expression, cannot access field null of object", object,
+          " (expression remainder: '", remainder , "'");
+      return undefined;
+    }
+    var value = (checkForThis && (fieldName == "this")) ? object : object[fieldName];
+    if (!remainder) return value;
+    if (remainder[0] == ".") {
+      return parseLiteralFieldAccess(value, remainder.slice(1, remainder.length), false);
```
Comment (shundread) [inline app/js/errantry/Game.js]: `this.parseLiteralFieldAccess(...)`

File: app/js/errantry/Game.js [hunk 5]
```diff
@@ -5038,6 +5048,61 @@ extObj.height = textObj.origHeight;
...
+    }
+    if (remainder[0] == "[") {
+      return parseBracketAccess(value, remainder);
```
Comment (shundread) [inline app/js/errantry/Game.js]: `this.parseBracketAccess(...)`

File: app/js/errantry/Game.js [hunk 6]
```diff
@@ -5038,6 +5048,61 @@ extObj.height = textObj.origHeight;
...
     return this.replaceStringsWithValues2(output);
   },
+  parseExpression: function (accessText) {
+    var asNumber = Number(accessText);
+    if (!isNaN(asNumber)) return asNumber;
+    return this.parseLiteralFieldAccess(this.gamestate, accessText, true);
+  },
+  parseFieldAccess: function (object, fieldName, remainder, checkForThis) {
+    if (!object) {
+      console.warning("While parsing expression, cannot access field", fieldName,
```
Comment (shundread) [inline app/js/errantry/Game.js]: Uncaught TypeError: console.warning is not a function.

There are a few other cases in the pull request as well.

File: app/js/errantry/Game.js [hunk 7]
```diff
@@ -5038,6 +5048,61 @@ extObj.height = textObj.origHeight;
...
+    for (; depth && bracketEnd < bracketRemainder.length; ++bracketEnd) {
+      if (bracketRemainder[bracketEnd] == "[") ++depth;
+      if (bracketRemainder[bracketEnd] == "]") --depth;
+    }
+    if (depth) {
+      console.warning("Invalid expression, unpaired '[' missing ']' (expression remainder '["
+          + remainder + "')");
+      return undefined;
+    }
+    return this.parseFieldAccess(object, parseExpression(bracketRemainder.slice(0, bracketEnd)),
```
Comment (shundread) [inline app/js/errantry/Game.js]: return this.parseFieldAccess(object, **_this._**parseExpression(bracketRemainder.slice(0, bracketEnd)),

**Comment by shundread** [comment]
Once syntax errors are fixed, the function crashes on the following event:

zetSetCurrentNpc($zet_npc1) -> _(sets currentNpc to zet_npc1)_
launchDialogue($14)
_Dialogue_Line:_ `My name is ${zetNpcs[currentNpc]name}` _(crash)_

With the following error:
```
Uncaught RangeError: Maximum call stack size exceeded

parseLiteralFieldAccess	@	Game.js:5081
parseExpression	@	Game.js:5054
parseBracketAccess	@	Game.js:5103
parseFieldAccess	@	Game.js:5073
parseLiteralFieldAccess	@	Game.js:5088
parseExpression	@	Game.js:5054
parseBracketAccess	@	Game.js:5103
parseFieldAccess	@	Game.js:5073
parseLiteralFieldAccess	@	Game.js:5088
parseExpression	@	Game.js:5054
```

There are no crashes if I use the following instead:
**_Dialogue_Line:_** `My name is ${zetNpcs.zet_npc1.name}`

File: app/js/errantry/Game.js (hunk 1 above)
Comment (iridiankin) [inline app/js/errantry/Game.js]: Number.isNaN is es6, isNaN is 'old': however unreliable. Replacing with different, more strict structure

**Pull request merged**